File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.java
class Solution {
	static int min_start;
	static int min_length;
	static int[] min_length_arr;
    public int[] solution(int[] sequence, int k) {
        int[] answer = {};
        min_length = Integer.MAX_VALUE;
        min_length_arr = new int[2];
        int start = 0;int end = 0; int sum = 0;
        // 투 포인터 종료 부분 잘 알아두기... 이렇게 종료하는게 맘이 편하다..
        while(start < sequence.length) {
        	if(sum > k || end == sequence.length) 
        		sum -= sequence[start++];
        	else {
        		if(end < sequence.length)
            		sum += sequence[end++];
        	}
        	if(sum == k) {
        		int dif = end - start;
                if(dif < min_length){
                    min_length = dif;
                    min_length_arr[0] = start;
                    min_length_arr[1] = end - 1;
                }else if(dif == min_length){
                    if(start < min_length_arr[0]){
                        min_length_arr[0] = start;
                        min_length_arr[1] = end - 1;
                    }
                }
                sum -= sequence[start++];
        	}
        }
        answer = min_length_arr;
        return answer;
    }
}

File: 프로그래머스/2/178870. 연속된 부분 수열의 합/연속된 부분 수열의 합.js
function solution(sequence, k) {
  var answer = [];
  var start = 0;
  var end = 0;
  var sum = 0;
  // 이런게 있다고 함.
  var len = Infinity;
  while (start < sequence.length) {
    if (sum < k && end < sequence.length) {
      sum += sequence[end++];
    } else if (sum >= k || end == sequence.length) {
      sum -= sequence[start++];
    }
    if (sum == k) {
      var dif = end - start - 1;
      if (len > dif) {
        len = dif;
        answer = [start, end -1];
      }
    } else if (len == dif && answer[0] > start) {
      answer = [start, end - 1];
    }
  }
  return answer;
}
